File: src/main/java/com/thesilentnights/easylogin/service/task/KickPlayer.java
package com.thesilentnights.easylogin.service.task;

import com.thesilentnights.easylogin.utils.LogUtil;
import net.minecraft.network.chat.Component;
import net.minecraft.server.level.ServerPlayer;

import java.util.UUID;

public class KickPlayer extends Task {

    private final ServerPlayer serverPlayer;
    private int delay;

    public KickPlayer(ServerPlayer serverPlayer, int delay) {
        this.serverPlayer = serverPlayer;
        this.delay = delay;
    }


    @Override
    public void execute() {
        LogUtil.logInfo(KickPlayer.class, "KickPlayer: " + serverPlayer.getDisplayName().getString());
        serverPlayer.connection.disconnect(Component.literal("You didn't login in time"));
    }

    @Override
    public int getTickDelay() {
        return this.delay;
    }

    @Override
    public void reduceTickDelay(int tickDelay) {
        this.delay -= tickDelay;
    }

    @Override
    public boolean shouldCancel(UUID uuid) {
        return uuid == serverPlayer.getUUID();
    }
}

File: src/main/kotlin/com/thesilentnights/easylogin/service/task/KickPlayer.kt
package com.thesilentnights.easylogin.service.task

import com.thesilentnights.easylogin.service.TaskService
import net.minecraft.server.level.ServerPlayer
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger

class KickPlayer(var serverPlayer: ServerPlayer, var delay: Long) : Task {
    var tickCount: Long = 0
    val log: Logger = LogManager.getLogger(KickPlayer::class.java)

    override fun tick() {
        this.tickCount++
        if (this.tickCount >= this.delay) {
            serverPlayer.disconnect()
            log.info("player {} has been kicked due to timeout in login", serverPlayer.gameProfile.name)
            TaskService.cancelTask(
                TaskService.generateTaskIdentifier(
                    serverPlayer.getUUID(),
                    TaskService.Suffix.TIMEOUT.name
                )
            )
        }
    }
}
